crates/transpiler/src/passes/alap_schedule_analysis.rs

```diff
@@ -0,0 +1,234 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2025
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use crate::TranspilerError;
+use hashbrown::HashMap;
+use pyo3::prelude::*;
+use pyo3::types::PyDict;
+use qiskit_circuit::dag_circuit::{DAGCircuit, Wire};
+use qiskit_circuit::dag_node::{DAGNode, DAGOpNode};
+use qiskit_circuit::operations::{OperationRef, StandardInstruction};
+use qiskit_circuit::{Clbit, Qubit};
+use rustworkx_core::petgraph::prelude::NodeIndex;
+
+// This enum takes care of the calculations when the durations are specified
+// either in seconds (as float) or in dt (as integers).
+#[derive(Debug, Clone)]
+pub enum TimeValue {
+    Float(f64),
+    Int(u64),
+}
+
+impl From<f64> for TimeValue {
+    fn from(value: f64) -> Self {
+        TimeValue::Float(value)
+    }
+}
+
+impl From<u64> for TimeValue {
+    fn from(value: u64) -> Self {
+        TimeValue::Int(value)
+    }
+}
+
+pub fn run_alap_schedule_analysis(
+    dag: &DAGCircuit,
+    clbit_write_latency: u64,
+    node_durations: HashMap<NodeIndex, TimeValue>,
+) -> PyResult<HashMap<NodeIndex, TimeValue>> {
+    if dag.qregs().len() != 1 || !dag.qregs_data().contains_key("q") {
+        return Err(TranspilerError::new_err(
+            "ALAP schedule runs on physical circuits only",
+        ));
+    }
+
+    let mut node_start_time: HashMap<NodeIndex, f64> = HashMap::new();
+    let mut idle_before: HashMap<Wire, f64> = HashMap::new();
```

Comment:
What's the guarantee that these types are always `float` and they don't match the same unit as the duration?